### ISSUE TITLE
Performance considerations of PhoneNumber::equals

### ISSUE THREAD

**Issue by stollr** [giggsey/libphonenumber-for-php#567]
I have three questions about the `equals` of the `PhoneNumber` class. Depending on the answer there may be some ways to improve its performance.

1\. Is there a reason why there is no short circuit for identical objects? Comparing those as the first thing could save some operations:

```php
    public function equals(PhoneNumber $other)
    {
        if ($this === $other) {
            return true;
        }
        // the rest
     }
```

2\. The comparison of each property looks very complicated, although the task seem to be very simply. Why not doing
```php 
        if ($this->getCountryCode() != $other->getCountryCode()) {
            return false;
        }
```
instead of this
```php 
        $sameCountry = $this->hasCountryCode() == $other->hasCountryCode() &&
            (!$this->hasCountryCode() || $this->getCountryCode() == $other->getCountryCode());
```
? Would it result in issues that I miss?

3\. Is there a reason why all the comparisons are operating on the has and get methods instead of the properties directly? Wouldn't it be more performant to directly access the properties of  `$this` and `$other`?

For example 
```php
$this->countryCode == $other->countryCode
// vs.
$this->getCountryCode() == $other->getCountryCode()
```

**Comment by giggsey** [comment]
The short answer is that the function was migrated straight over from the Java upstream of the project.

I'd be happy to accept a merge request with all of your suggestions.

**Comment by giggsey** [comment]
Closed by #568
